flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/AdaptedRestartPipelinedRegionStrategyNG.java

```diff
@@ -0,0 +1,296 @@
+		// record current versions
+		final long globalModVersion = executionGraph.getGlobalModVersion();
+		final Set<ExecutionVertexVersion> vertexVersions = new HashSet<>(
+			executionVertexVersioner.recordVertexModifications(verticesToRestart).values());
+
+		// cancel tasks involved in this task failure
+		cancelTasks(verticesToRestart)
+			.thenAccept(
+				(Object ignored) -> {
+					// skip local failover if is in global failover
+					if (!isLocalFailoverValid(globalModVersion)) {
+						LOG.info("Skip current region failover as a global failover is ongoing.");
+						return;
+					}
+
+					// found out vertices which are still valid to restart.
+					// some vertices involved in this failover may be modified if another region
+					// failover happens during the cancellation stage of this failover.
+					// Will ignore the modified vertices as the other failover will deal with them.
+					final Set<ExecutionVertex> unmodifiedVertices = executionVertexVersioner
+						.getUnmodifiedExecutionVertices(vertexVersions)
+						.stream()
+						.map(this::getExecutionVertex)
+						.collect(Collectors.toSet());
+
+					try {
+						LOG.info("Finally restart {} tasks to recover from task failure.", unmodifiedVertices.size());
+
+						// reset tasks to CREATED state and reload state
+						resetTasks(unmodifiedVertices, globalModVersion);
+
+						// re-schedule tasks
+						rescheduleTasks(unmodifiedVertices, globalModVersion);
+					} catch (GlobalModVersionMismatch e) {
+						// happens when a global recovery happens concurrently to the regional recovery
+						// just stop this local failover
+					} catch (Exception e) {
+						throw new CompletionException(e);
+					}
+				})
+			.whenComplete(
+				(Object ignored, Throwable t) -> {
+					// fail globally if any error happens
+					if (t != null) {
+						LOG.info("Unexpected error happens in region failover. Fail globally.", t);
+						failGlobal(t);
+					}
+				});
+	}
+
+	@VisibleForTesting
+	protected CompletableFuture<?> cancelTasks(final Set<ExecutionVertexID> vertices) {
+		final List<CompletableFuture<?>> cancelFutures = vertices.stream()
+			.map(this::cancelExecutionVertex)
+			.collect(Collectors.toList());
+
+		return FutureUtils.combineAll(cancelFutures);
+	}
+
+	private void resetTasks(final Set<ExecutionVertex> vertices, final long globalModVersion) throws Exception {
+		final Set<CoLocationGroup> colGroups = new HashSet<>();
+		final long restartTimestamp = System.currentTimeMillis();
+
+		for (ExecutionVertex ev : vertices) {
+			CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
+			if (cgroup != null && !colGroups.contains(cgroup)){
+				cgroup.resetConstraints();
+				colGroups.add(cgroup);
+			}
+
+			ev.resetForNewExecution(restartTimestamp, globalModVersion);
+		}
+
+		// if there is checkpointed state, reload it into the executions
+		if (executionGraph.getCheckpointCoordinator() != null) {
```

Comment:
What are the chances to move this into `resetForNewExecution()`? Now this code is duplicated.

Reply:
This is not ExecutionVertex level operation so it cannot be merged into ExecutionVertex::resetForNewExecution.
If you are really concerning about the code duplication of partial checkpoint restore between the new strategy and the legacy strategy, I can merge them.

IMO, we keep the legacy RestartPipelinedStrategy in case that the new strategy has some bugs. Making changes to legacy RestartPipelinedStrategy is adding bug risk to it. I think it would be better to leave it as is, since it is actually in deprecated state.